### PR TITLE
docs: add documenting guidelines from wiki

### PR DESCRIPTION
**Pull request by clason**

### CODE REVIEW

**Review by lewis6991 — APPROVED** [review]

**Review by vigoux — APPROVED** [review]

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -266,6 +266,41 @@ For managing includes in C files, use [include-what-you-use].
...
+--- {Long explanation}
+---
+---@param arg1 type {description}
+---@param arg2 type {description}
+{...}
+---
+---@return type {description}
+```
+
+If possible, always add type information (`table`, `string`, `number`, ...). Multiple valid types are separated by a bar (`string|table`). Indicate optional parameters via `type|nil`.
```
[the remark this answers is not in the document]
Comment (lewis6991) [reply CONTRIBUTING.md:291]: I think longer term we should fully adopt emmylua. Right now we're using some rough partial subset inconsistently.

[the remark this answers is not in the document]
Comment (clason) [reply CONTRIBUTING.md:291]: We do; we just don't make use of every arrow in its quiver (nor do I think we should). I also prefer `arg1 (type) description` over `arg1 type description`, which both `lua-language-server` and `vim_docgen.py` can handle.

The fact that we're inconsistent is an issue, of course, that I'll hope will be addressed incrementally (either by encouraging people to annotate every function they touch, or by dedicated annotation PRs).

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -266,6 +266,41 @@ For managing includes in C files, use [include-what-you-use].
 
 See [#549][549] for more details.
 
+Documenting
```
Comment (justinmk) [inline CONTRIBUTING.md:269]: isn't this redundant with `:help dev-doc` ? idk how this found its way to the wiki

Comment (clason) [reply CONTRIBUTING.md:269]: No objection to merging this with `dev-doc`; the separation feels arbitrary anyway. (Personally, I find the markdown files easier to read, especially when they make use of links to web pages.)

**Pull request merged**